### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

In [2]:
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [3]:
#check for imcomplete rows
purchase_data.count()

Purchase ID    780
SN             780
Age            780
Gender         780
Item ID        780
Item Name      780
Price          780
dtype: int64

In [4]:
#check the data types in the dataframe
purchase_data.dtypes

Purchase ID      int64
SN              object
Age              int64
Gender          object
Item ID          int64
Item Name       object
Price          float64
dtype: object

## Player Count

* Display the total number of players


In [17]:
#Use len to return no. of items in the array
total_players = len(purchase_data["SN"].unique())
total_players

576

In [18]:
#Present data in a dataframe
total_players_df = pd.DataFrame({"Total Players": [total_players]})
total_players_df

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [19]:
#Use len to return no. of unique items in the array
total_items = len(purchase_data["Item ID"].unique())
total_items

#Calculate the average purchase price
average_price = purchase_data["Price"].mean()
average_price

#Calulate the Total number of Purchases
total_purchases = purchase_data["Purchase ID"].count()
total_purchases

#Calculate the Total Revenue
total_revenue = purchase_data["Price"].sum()
total_revenue

# Create a dataframe to display the summary results
purchase_summary_df = pd.DataFrame(
                            {"Number of Unique Items": [total_items],
                             "Average Price": [average_price],
                             "Number of Purchases" : [total_purchases],
                             "Total Revenue" : [total_revenue] }
                                                    )
purchase_summary_df

Number of Unique Items  Average Price  Number of Purchases  Total Revenue
0                     179       3.050987                  780        2379.77

In [20]:
#Format the average price and revenue to 2 decimal places
purchase_summary_df["Average Price"] = purchase_summary_df["Average Price"].map("{:,.2f}%".format)
purchase_summary_df["Total Revenue"] = purchase_summary_df["Total Revenue"].map("${:,.2f}".format)
purchase_summary_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         3.05%                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [21]:
#Drop duplicate rows of player id to get unique rows in dataframe
gender_df = purchase_data.drop_duplicates(subset=['SN'])
gender_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [22]:
#Use value_counts to get number of rows
gender_df["SN"].value_counts()

Leetirraya83    1
Lisotesta51     1
Yasrisu92       1
Chadossa89      1
Ali84           1
               ..
Pheodai94       1
Quanenrian83    1
Phyali88        1
Lassimla92      1
Lirtim36        1
Name: SN, Length: 576, dtype: int64

In [26]:
# Group by Gender
grouped_gender = gender_df.groupby(["Gender"]).count()

# Sort by highest value
grouped_gender = grouped_gender.sort_values("SN", ascending = False)

# Reduce the columns in dataframe
grouped_gender = grouped_gender[["SN"]]

# Rename the column header
grouped_gender = grouped_gender.rename(columns = {"SN": "Total Count"})
grouped_gender

Total Count
Gender                            
Male                           484
Female                          81
Other / Non-Disclosed           11

In [27]:
# Calculate Percentage of Players
# Access each row in dataframe, apply formula to data in each row and write result to new column
grouped_gender["Percentage of Players"] = (grouped_gender['Total Count'] / total_players) * 100 

#Format the 2 decimal places and add percentage sign
grouped_gender["Percentage of Players"] = grouped_gender["Percentage of Players"].map("{:,.2f}%".format)
grouped_gender

Total Count Percentage of Players
Gender                                                  
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [29]:
# Count the number of players by gender
purchase_count = purchase_data.groupby(["Gender"]).count()

# Reduce the columns in dataframe
purchase_count = purchase_count[["Purchase ID"]]

# Rename the column to Purchase Count
purchase_count = purchase_count.rename(columns = {"Purchase ID": "Purchase Count"})
purchase_count

Purchase Count
Gender                               
Female                            113
Male                              652
Other / Non-Disclosed              15

In [30]:
# Get the average price purchased by gender
average_price = purchase_data.groupby(["Gender"]).mean()

# Reduce the columns in dataframe
average_price = average_price[["Price"]]

#Format the price to 2 decimal places and add percentage sign
average_price["Price"] = average_price["Price"].map("${:,.2f}".format)

# Rename the column to Average purchase price
average_price = average_price.rename(columns = {"Price": "Average Purchase Price"})
average_price

Average Purchase Price
Gender                                      
Female                                 $3.20
Male                                   $3.02
Other / Non-Disclosed                  $3.35

In [32]:
#Get the total purchase price by gender
total_purchase = purchase_data.groupby(["Gender"]).sum()

# Reduce the columns in dataframe
total_purchase = total_purchase[["Price"]]

# Rename the column to Total purchase price
total_purchase = total_purchase.rename(columns = {"Price": "Total Purchase Price"})
total_purchase

Total Purchase Price
Gender                                     
Female                               361.94
Male                                1967.64
Other / Non-Disclosed                 50.19

In [33]:
# Calculate Avg Total Purchase per Person
Avg_total_purchase = grouped_gender
Avg_total_purchase["Avg Total Purchase per Person"] = total_purchase["Total Purchase Price"] /grouped_gender["Total Count"] 

# Reduce the columns in dataframe
Avg_total_purchase = Avg_total_purchase[["Avg Total Purchase per Person"]]
Avg_total_purchase 

Avg Total Purchase per Person
Gender                                              
Male                                        4.065372
Female                                      4.468395
Other / Non-Disclosed                       4.562727

In [34]:
# Merge the dataframes and assign the results to summary dataframe
purchasing_summary = pd.merge(purchase_count, average_price, on="Gender")
purchasing_summary = pd.merge(purchasing_summary, total_purchase, on="Gender")
purchasing_summary = pd.merge(purchasing_summary, Avg_total_purchase , on="Gender")

#Format the amounts to 2 decimal places and add dollar/percentage sign
purchasing_summary["Total Purchase Price"] = purchasing_summary["Total Purchase Price"].map("${:,.2f}".format)
purchasing_summary["Avg Total Purchase per Person"] = purchasing_summary["Avg Total Purchase per Person"].map("${:,.2f}".format)
purchasing_summary

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Price Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [36]:
# Create bins for ages
bins = [0, 10, 15, 20, 25, 30, 35, 40, 50]

# Create labels for the bins
age_labels = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]
print(age_labels)

['<10', '10-14', '15-19', '20-24', '25-29', '30-34', '35-39', '40+']


In [38]:
#Drop duplicate rows of player id to get unique rows in dataframe
age_demographics = purchase_data.drop_duplicates(subset=['SN'])
age_demographics.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [46]:
#Bin the purchase data by Age and put the results in dataframe
age_demographics["Age Group"] = pd.cut(age_demographics["Age"], bins, labels=age_labels, right=False)  

C:\Users\ianli\anaconda3\envs\PythonData\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [47]:
# Group the data by Age Group & Count the number of players in age group
age_group = age_demographics.groupby(["Age Group"]).count()
age_group.head()

Purchase ID   SN  Age  Gender  Item ID  Item Name  Price
Age Group                                                          
<10                 17   17   17      17       17         17     17
10-14               22   22   22      22       22         22     22
15-19              107  107  107     107      107        107    107
20-24              258  258  258     258      258        258    258
25-29               77   77   77      77       77         77     77

In [48]:
# Reduce the columns in dataframe
age_group = age_group[["SN"]]
age_group.head()

SN
Age Group     
<10         17
10-14       22
15-19      107
20-24      258
25-29       77

In [49]:
# Calculate Percentage of Players
# Access each row in dataframe, apply formula to data in each row and write result to new column
age_group["Percentage of Players"] = (age_group['SN'] / total_players) * 100 

#Format the percentage of players to 2 decimal places and add percentage sign
age_group["Percentage of Players"] = age_group["Percentage of Players"].map("{:,.2f}%".format)

# Rename the column title
age_group = age_group.rename(columns = {"SN": "Total Count"})
age_group

Total Count Percentage of Players
Age Group                                   
<10                 17                 2.95%
10-14               22                 3.82%
15-19              107                18.58%
20-24              258                44.79%
25-29               77                13.37%
30-34               52                 9.03%
35-39               31                 5.38%
40+                 12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [50]:
# Create bins for ages
bins = [0, 10, 15, 20, 25, 30, 35, 40, 50]

# Create labels for the bins
age_labels = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]
print(age_labels)

['<10', '10-14', '15-19', '20-24', '25-29', '30-34', '35-39', '40+']


In [51]:
#Bin the purchase data by Age and put the results in dataframe
purchase_demographics = purchase_data
purchase_demographics["Age Ranges"] = pd.cut(purchase_data["Age"], bins, labels=age_labels, right=False)
purchase_demographics

Purchase ID             SN  Age  Gender  Item ID  \
0              0        Lisim78   20    Male      108   
1              1    Lisovynya38   40    Male      143   
2              2     Ithergue48   24    Male       92   
3              3  Chamassasya86   24    Male      100   
4              4      Iskosia90   23    Male      131   
..           ...            ...  ...     ...      ...   
775          775     Aethedru70   21  Female       60   
776          776         Iral74   21    Male      164   
777          777     Yathecal72   20    Male       67   
778          778        Sisur91    7    Male       92   
779          779      Ennrian78   24    Male       50   

                                     Item Name  Price Age Ranges  
0    Extraction, Quickblade Of Trembling Hands   3.53      20-24  
1                            Frenzied Scimitar   1.56        40+  
2                                 Final Critic   4.88      20-24  
3                                  Blindscythe   3.27      20-24  
4                                         Fury   1.44      20-24  
..                                         ...    ...        ...  
775                                       Wolf   3.54      20-24  
776                           Exiled Doomblade   1.63      20-24  
777      Celeste, Incarnation of the Corrupted   3.46      20-24  
778                               Final Critic   4.19        <10  
779                                       Dawn   4.60      20-24  

[780 rows x 8 columns]

In [55]:
# Group the data by Age Group & Count the number of players in age group
purchase_age_df = purchase_demographics.groupby(["Age Ranges"]).count()

# Reduce the columns in dataframe
purchase_age_count = purchase_age_df[["Purchase ID"]]

# Rename the column to Purchase Count
purchase_age_count = purchase_age_count.rename(columns = {"Purchase ID": "Purchase Count"})
purchase_age_count

Purchase Count
Age Ranges                
<10                     23
10-14                   28
15-19                  136
20-24                  365
25-29                  101
30-34                   73
35-39                   41
40+                     13

In [56]:
# Group the data by Age Group & Count the number of players in age group
purchase_average_df = purchase_demographics.groupby(["Age Ranges"]).mean()

# Rename the column to Average purchase price
purchase_average_df = purchase_average_df.rename(columns = {"Price": "Average Purchase Price"})

# Reduce the columns in dataframe
purchase_average_df = purchase_average_df[["Average Purchase Price"]]

# #Format the price to 2 decimal places and add percentage sign
purchase_average_df["Average Purchase Price"] = purchase_average_df["Average Purchase Price"].map("${:,.2f}".format)
purchase_average_df

Average Purchase Price
Age Ranges                       
<10                         $3.35
10-14                       $2.96
15-19                       $3.04
20-24                       $3.05
25-29                       $2.90
30-34                       $2.93
35-39                       $3.60
40+                         $2.94

In [57]:
#Total purchase price by gender
total_purchase_df = purchase_demographics.groupby(["Age Ranges"]).sum()
total_purchase_df

# Reduce the columns in dataframe
total_purchase_df = total_purchase_df[["Price"]]

#Rename the column to Total purchase value
total_purchase_df = total_purchase_df.rename(columns = {"Price": "Total Purchase Value"})
total_purchase_df

Total Purchase Value
Age Ranges                      
<10                        77.13
10-14                      82.78
15-19                     412.89
20-24                    1114.06
25-29                     293.00
30-34                     214.00
35-39                     147.67
40+                        38.24

In [60]:
#Drop duplicate rows of player id to get unique rows in dataframe
purchase_age_unique_count = purchase_data.drop_duplicates(subset=['SN'])
purchase_age_unique_count.head()

# Group the data by Age Group & Count the number of players in age group
purchase_age_unique_count = purchase_age_unique_count.groupby(["Age Ranges"]).count()
purchase_age_unique_count

Purchase ID   SN  Age  Gender  Item ID  Item Name  Price
Age Ranges                                                          
<10                  17   17   17      17       17         17     17
10-14                22   22   22      22       22         22     22
15-19               107  107  107     107      107        107    107
20-24               258  258  258     258      258        258    258
25-29                77   77   77      77       77         77     77
30-34                52   52   52      52       52         52     52
35-39                31   31   31      31       31         31     31
40+                  12   12   12      12       12         12     12

In [62]:
# Calculate Avg Total Purchase per Person
Avg_total_purchase_person = purchase_age_unique_count
Avg_total_purchase_person["Avg Total Purchase per Person"] = total_purchase_df["Total Purchase Value"] /purchase_age_unique_count["SN"] 

# Reduce the columns in dataframe
Avg_total_purchase_person = Avg_total_purchase_person[["Avg Total Purchase per Person"]]
Avg_total_purchase_person

Avg Total Purchase per Person
Age Ranges                               
<10                              4.537059
10-14                            3.762727
15-19                            3.858785
20-24                            4.318062
25-29                            3.805195
30-34                            4.115385
35-39                            4.763548
40+                              3.186667

In [63]:
# Merge the dataframes and assign the results to summary dataframe
purchasing_age_summary = pd.merge(purchase_age_count, purchase_average_df, on="Age Ranges")
purchasing_age_summary = pd.merge(purchasing_age_summary, total_purchase_df , on="Age Ranges")
purchasing_age_summary = pd.merge(purchasing_age_summary, Avg_total_purchase_person , on="Age Ranges")

#Format the 2 decimal places and add dollar/percentage sign
purchasing_age_summary["Total Purchase Value"] = purchasing_age_summary["Total Purchase Value"].map("${:,.2f}".format)
purchasing_age_summary["Avg Total Purchase per Person"] = purchasing_age_summary["Avg Total Purchase per Person"].map("${:,.2f}".format)
purchasing_age_summary

Purchase Count Average Purchase Price Total Purchase Value  \
Age Ranges                                                               
<10                     23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05            $1,114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
40+                     13                  $2.94               $38.24   

           Avg Total Purchase per Person  
Age Ranges                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [64]:
# Group the data by Player ID & aggregate the columns by Count 
spenders_df = purchase_data.groupby(["SN"]).count()
spenders_df

# Reduce the columns in dataframe
spenders_df = spenders_df[["Purchase ID"]]

# # Rename the column to Total purchase value
spenders_df = spenders_df.rename(columns = {"Purchase ID": "Purchase Count"})
spenders_df

Purchase ID  Age  Gender  Item ID  Item Name  Price  Age Ranges
SN                                                                            
Adairialis76             1    1       1        1          1      1           1
Adastirin33              1    1       1        1          1      1           1
Aeda94                   1    1       1        1          1      1           1
Aela59                   1    1       1        1          1      1           1
Aelaria33                1    1       1        1          1      1           1
...                    ...  ...     ...      ...        ...    ...         ...
Yathecal82               3    3       3        3          3      3           3
Yathedeu43               2    2       2        2          2      2           2
Yoishirrala98            1    1       1        1          1      1           1
Zhisrisu83               2    2       2        2          2      2           2
Zontibe81                3    3       3        3          3      3           3

[576 rows x 7 columns]

In [67]:
# Group the data by Player ID & aggregate the columns by average
spenders_avg_df = purchase_data.groupby(["SN"]).mean()

# Reduce the columns in dataframe
spenders_avg_df = spenders_avg_df[["Price"]]

# # Rename the column to Total purchase value
spenders_avg_df = spenders_avg_df.rename(columns = {"Price": "Average Purchase Price"})
spenders_avg_df

Average Purchase Price
SN                                   
Adairialis76                 2.280000
Adastirin33                  4.480000
Aeda94                       4.910000
Aela59                       4.320000
Aelaria33                    1.790000
...                               ...
Yathecal82                   2.073333
Yathedeu43                   3.010000
Yoishirrala98                4.580000
Zhisrisu83                   3.945000
Zontibe81                    2.676667

[576 rows x 1 columns]

In [69]:
# Group the data by Player ID & aggregate the columns by total
spenders_total_df = purchase_data.groupby(["SN"]).sum()

# Reduce the columns in dataframe
spenders_total_df = spenders_total_df[["Price"]]

# # Rename the column to Total purchase value
spenders_total_df = spenders_total_df.rename(columns = {"Price": "Total Purchase Value"})
spenders_total_df

Total Purchase Value
SN                                 
Adairialis76                   2.28
Adastirin33                    4.48
Aeda94                         4.91
Aela59                         4.32
Aelaria33                      1.79
...                             ...
Yathecal82                     6.22
Yathedeu43                     6.02
Yoishirrala98                  4.58
Zhisrisu83                     7.89
Zontibe81                      8.03

[576 rows x 1 columns]

In [71]:
# Merge the dataframes and assign the results to summary dataframe
top_spenders_summary = pd.merge(spenders_df, spenders_avg_df, on="SN")
top_spenders_summary = pd.merge(top_spenders_summary, spenders_total_df , on="SN")

# Sort the total purchase value in descending order
top_spenders_summary = top_spenders_summary.sort_values("Total Purchase Value", ascending=False)

#Format the 2 decimal places and add dollar/percentage sign
top_spenders_summary["Average Purchase Price"] = top_spenders_summary["Average Purchase Price"].map("${:,.2f}".format)
top_spenders_summary["Total Purchase Value"] = top_spenders_summary["Total Purchase Value"].map("${:,.2f}".format)

# Display a preview of the summary dataframe
top_spenders_summary.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [72]:
# Reduce the colomns of the dataframe to Item ID, Item Name & Price
popular_items = purchase_data[["Item ID", "Item Name", "Price"]]
popular_items

Item ID                                  Item Name  Price
0        108  Extraction, Quickblade Of Trembling Hands   3.53
1        143                          Frenzied Scimitar   1.56
2         92                               Final Critic   4.88
3        100                                Blindscythe   3.27
4        131                                       Fury   1.44
..       ...                                        ...    ...
775       60                                       Wolf   3.54
776      164                           Exiled Doomblade   1.63
777       67      Celeste, Incarnation of the Corrupted   3.46
778       92                               Final Critic   4.19
779       50                                       Dawn   4.60

[780 rows x 3 columns]

In [73]:
# Group the dataframe by Item ID and Item Name and calculate count
popular_itemscount = popular_items.groupby(["Item ID", "Item Name"]).count()

# Rename the column in dataframe
popular_itemscount = popular_itemscount.rename(columns = {"Price": "Purchase Count"})
popular_itemscount

Purchase Count
Item ID Item Name                                                   
0       Splinter                                                   4
1       Crucifer                                                   4
2       Verdict                                                    6
3       Phantomlight                                               6
4       Bloodlord's Fetish                                         5
...                                                              ...
178     Oathbreaker, Last Hope of the Breaking Storm              12
179     Wolf, Promise of the Moonwalker                            6
181     Reaper's Toll                                              5
182     Toothpick                                                  3
183     Dragon's Greatsword                                        3

[179 rows x 1 columns]

In [74]:
# Calculate the average for the dataframe
popular_items_average = popular_items.groupby(["Item ID", "Item Name"]).mean()

# Rename the column in dataframe
popular_items_average = popular_items_average.rename(columns = {"Price": "Item Price"})
popular_items_average

Item Price
Item ID Item Name                                               
0       Splinter                                          1.2800
1       Crucifer                                          2.9425
2       Verdict                                           2.4800
3       Phantomlight                                      2.4900
4       Bloodlord's Fetish                                1.7000
...                                                          ...
178     Oathbreaker, Last Hope of the Breaking Storm      4.2300
179     Wolf, Promise of the Moonwalker                   4.4800
181     Reaper's Toll                                     1.6600
182     Toothpick                                         4.0300
183     Dragon's Greatsword                               1.0900

[179 rows x 1 columns]

In [75]:
# Calculate the total for the dataframe
popular_items_total = popular_items.groupby(["Item ID", "Item Name"]).sum()

# Rename the column in dataframe
popular_items_total = popular_items_total.rename(columns = {"Price": "Total Purchase Value"})
popular_items_total

Total Purchase Value
Item ID Item Name                                                         
0       Splinter                                                      5.12
1       Crucifer                                                     11.77
2       Verdict                                                      14.88
3       Phantomlight                                                 14.94
4       Bloodlord's Fetish                                            8.50
...                                                                    ...
178     Oathbreaker, Last Hope of the Breaking Storm                 50.76
179     Wolf, Promise of the Moonwalker                              26.88
181     Reaper's Toll                                                 8.30
182     Toothpick                                                    12.09
183     Dragon's Greatsword                                           3.27

[179 rows x 1 columns]

In [76]:
#Merge tables to add item price and place results in summary dataframe
popular_items_summary = pd.merge(popular_itemscount, popular_items_average, on=["Item ID","Item Name"])
popular_items_summary = pd.merge(popular_items_summary, popular_items_total, on=["Item ID","Item Name"])

#Sort the purchase count column in descending order
popular_items_summary = popular_items_summary.sort_values("Purchase Count", ascending=False)

#Format the 2 decimal places and add dollar/percentage sign
popular_items_summary["Item Price"] = popular_items_summary["Item Price"].map("${:,.2f}".format)
popular_items_summary["Total Purchase Value"] = popular_items_summary["Total Purchase Value"].map("${:,.2f}".format)

# Display a preview of summary dataframe
popular_items_summary.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [77]:
# Reduce the colomns of the dataframe to Item ID, Item Name & Price
profitable_items = purchase_data[["Item ID", "Item Name", "Price"]]
profitable_items

Item ID                                  Item Name  Price
0        108  Extraction, Quickblade Of Trembling Hands   3.53
1        143                          Frenzied Scimitar   1.56
2         92                               Final Critic   4.88
3        100                                Blindscythe   3.27
4        131                                       Fury   1.44
..       ...                                        ...    ...
775       60                                       Wolf   3.54
776      164                           Exiled Doomblade   1.63
777       67      Celeste, Incarnation of the Corrupted   3.46
778       92                               Final Critic   4.19
779       50                                       Dawn   4.60

[780 rows x 3 columns]

In [78]:
# Group the dataframe by Item ID and Item Name and calculate count
profitable_itemscount = profitable_items.groupby(["Item ID", "Item Name"]).count()

# Rename the column in dataframe
profitable_itemscount = profitable_itemscount.rename(columns = {"Price": "Purchase Count"})
profitable_itemscount

Purchase Count
Item ID Item Name                                                   
0       Splinter                                                   4
1       Crucifer                                                   4
2       Verdict                                                    6
3       Phantomlight                                               6
4       Bloodlord's Fetish                                         5
...                                                              ...
178     Oathbreaker, Last Hope of the Breaking Storm              12
179     Wolf, Promise of the Moonwalker                            6
181     Reaper's Toll                                              5
182     Toothpick                                                  3
183     Dragon's Greatsword                                        3

[179 rows x 1 columns]

In [79]:
# Calculate the average for the dataframe
profitable_items_average = profitable_items.groupby(["Item ID", "Item Name"]).mean()

# Rename the column in dataframe
profitable_items_average = profitable_items_average.rename(columns = {"Price": "Item Price"})
profitable_items_average

Item Price
Item ID Item Name                                               
0       Splinter                                          1.2800
1       Crucifer                                          2.9425
2       Verdict                                           2.4800
3       Phantomlight                                      2.4900
4       Bloodlord's Fetish                                1.7000
...                                                          ...
178     Oathbreaker, Last Hope of the Breaking Storm      4.2300
179     Wolf, Promise of the Moonwalker                   4.4800
181     Reaper's Toll                                     1.6600
182     Toothpick                                         4.0300
183     Dragon's Greatsword                               1.0900

[179 rows x 1 columns]

In [80]:
# Calculate the total for the dataframe
profitable_items_total = popular_items.groupby(["Item ID", "Item Name"]).sum()

# Rename the column in dataframe
profitable_items_total = popular_items_total.rename(columns = {"Price": "Total Purchase Value"})
profitable_items_total

Total Purchase Value
Item ID Item Name                                                         
0       Splinter                                                      5.12
1       Crucifer                                                     11.77
2       Verdict                                                      14.88
3       Phantomlight                                                 14.94
4       Bloodlord's Fetish                                            8.50
...                                                                    ...
178     Oathbreaker, Last Hope of the Breaking Storm                 50.76
179     Wolf, Promise of the Moonwalker                              26.88
181     Reaper's Toll                                                 8.30
182     Toothpick                                                    12.09
183     Dragon's Greatsword                                           3.27

[179 rows x 1 columns]

In [81]:
#Merge tables to add item price
profitable_items_summary = pd.merge(profitable_itemscount, profitable_items_average, on=["Item ID","Item Name"])
profitable_items_summary = pd.merge(profitable_items_summary, profitable_items_total, on=["Item ID","Item Name"])

#Sort the purchase count column in descending order
profitable_items_summary = profitable_items_summary.sort_values("Total Purchase Value", ascending=False)

#Format the 2 decimal places and add dollar/percentage sign
profitable_items_summary["Item Price"] = profitable_items_summary["Item Price"].map("${:,.2f}".format)
profitable_items_summary["Total Purchase Value"] = profitable_items_summary["Total Purchase Value"].map("${:,.2f}".format)
profitable_items_summary.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

Observations:

1.  There are significantly more male players than female players with male players representing 84% of the players in the game.
    This game is more popular with male players than female players.

2.  The top 3 groups of players fall in the age range of 15-29 years old making up 76% of the players.  The most players are in the age range of 20-24 years old representing 44.79% of the players.

3.  The best selling item is Final Critic with the most purchases and is the most profitable items purchased by players.  This is closely followed by the next best selling item Oathbreaker, Last Hope of the Breaking Storm.
